Possible SMW v2 implementation

TheLX5 opened the pull request:
For the past half year I've been working on upgrading the SMW apworld for a close friend group to make it even more interesting than it already is, I added several features which are listed below to create even more varied SMW sessions in AP.

## New features
* Blocksanity
  * Adds checks for almost every block containing an item inside (657 blocks) which include:
    * Question Mark blocks
    * Invisible blocks
    * Turn blocks
    * Note blocks
    * Flying Question Mark blocks
    * Switch Palace blocks
    * Bonus blocks (green blocks with a star)
  * Excludes the following blocks from being added:
    * Blocks in Front Door
      * Makes no sense for a player to enter Front Door if they're playing Yoshi Egg Hunt
    * Top Secret Area blocks
      * Couldn't figure out how to add them on the Python side (lol)
    * Blocks in not casual friendly areas
      * Most of the Gnarly Question Mark blocks in the sublevel except the ones used for a stair
      * The Switch Palace blocks bridge in one of the CI2 sublevels
      * Some Yellow Switch Palace blocks in YI3 that require the player to fly below ground
      * May be some more
* Checks for 3-up Moons
  * Adds 7 extra checks 
  * Even though they're not that interesting to grab as most of them require some sort of flying, adding checks for the sake of adding them seemed fun
* Checks for getting a 1-up from a Bonus block
  * There are only 4 checks, but they're an extra challenge to the levels they're in so I added them anyway
  * They're somewhat casual friendly checks unless the person doesn't know you need to collect 30 coins in the level without dying to get the prize
* Checks for Invisible 1-up Checkpoints
  * Adds 14 checks 
  * These checks are for people who are well versed in SMW. They're located in very reachable areas, but since they're invisible, you can't know where are they
  * Definitely not a casual friendly option (might make a infographic later...)
* Curated random palettes
  * Adds an extensive selection of palettes which are added into the game with the goal of having better palettes
  * All of them were made by me and my friends. Most of them were properly reviewed to ensure they're neat to look at. Wouldn't hurt to have a second opinion on this one though.
* Shuffled SFXs
  * Sound effect calls are now shuffled, everything is random and there's no sort of control to know where it places new SFXs
  * Have fun with wacky combinations :D  
* In-level indicators
  * Added small images during level gameplay ([YouTube](https://www.youtube.com/watch?v=EbQm2SC4g6g)) that are shown whenever the player receives a junk item, a Yoshi Egg or a Boss Token.
* Map Indicators
  * Added indicators ([YouTube](https://youtu.be/ljQ_WBNs3ac)) in the Map Border that tell you which upgrades you've received and the checks you've done in a level (exits, dragon coins, blocksanity, etc)
  * Collected upgrades and checks are marked with a !, uncollected upgrades and checks are marked with a dot.
  * In order:
    * (Po) Progressive Powerup (0-3)
    * (Sp) Spinjump
    * (Ru) Run
    * (Ca) Carry
    * (Sw) Swim
    * (Cl) Climb
    * (Yo) Yoshi
    * (Ps) P-Switch
    * (Pb) P-Balloon
    * (St) Star Activate
    * (YS) Yellow Switch Palace
    * (GS) Green Switch Palace
    * (RS) Red Switch Palace
    * (BS) Blue Switch Palace
    * (Sp) Special World Clear
    * (E1) Exit 1/Normal Exit
    * (E2) Exit 2/Secret Exit/Key
    * (DC) All Dragon Coins collected
    * (Mo) Moon collected
    * (Ch) 1-up Checkpoints collected
    * (Bo) Prize from Bonus block collected
    * (BS) All blocks collected
* Added a "Special World Clear" item
  * Necessary for blocksanity
  * It's also a cool addition to make some levels much easier when having a Yoshi
* Added extra junk items
  * They're all coins (+1, +5, +10, + 15)
  * Added so you just don't receive 1-ups from the extra checks that Blocksanity adds
  * The coins don't work towards the Bonus block coin count
* Added the possibility of swapping the player character with a custom one
  * The implementation decompresses the player graphics and puts them in a unused area alongside with its hands found elsewhere in the game. This makes much easier to swap out the data with your own stuff.
  * Making it via decompression ensures that no copyrighted data is bundled with SMW v2's apworld
  * Easier swapping requires an adjuster, but my Python-fu is very lacking to do one of these
  * BTW this also applies to Yoshi and every block that's animated via VRAM writes
 
## Edits
* Made the camera's vertical scroll edit optional
  * It caused some levels to have blind jumps
* Moved Butter Bridge 2 to the hard level list
  * I have the power to do it (I have skill issue)
* Made Awesome require a Fire Flower
  *  Same reason as above
* Adjusted Funky upgrade requirements
  * It only needed a progressive powerup, I added extra ways to mark it as beatable

## Side effects
* The patched game results in a 1MiB file instead of a 512KiB one
  * There wasn't enough free data for Blocksanity tables (needed over 0x1000 consecutive bytes)
  * This data is now used by palettes and other enhancements
* The save file is now 4KiB instead of 2KiB
  * This one is funny, I couldn't reuse the 2KiB area for my data because it corrupted itself, so I decided to pull a "DK64 dev" and expanded the RAM, that solved the corruption issue.
* The received items RAM is now 16-bits. Makes the corresponding lines in the Client script a bit weird.
  *  Maybe there's a way to handle 16-bit reads in a more elegant way

## More talk
I apologize in advance for the quality of the Python code. I'm not a competent Python programmer, my work was based on what was already there and a bunch of Google queries. I'm more adept at writing SNES ASM and I know SMW and its internals very well to handle creating the new features with minimal bugs in the process.
Feel free to request to me to rewrite some stuff, but I wouldn't bother you if you decide to refactor my code to actually fit into your standards or Archipelago's standards.

Blocksanity python code (regions, locations, etc) and tables and my ASM code in Python were generated with Python scripts. I was too lazy to adapt them by hand, especially that many lines lol @ blocksanity. My ASM code was compiled with [Asar v1.81](https://github.com/RPGHacker/asar). The palettes were created with [Lunar Magic v3.30](https://fusoya.eludevisibility.org/lm/program.html).

TheLX5 committed:
Initial v2 files

they work on AP 0.4.3

PoryGone commented:
Thank you for the rigourous Changelist, that's very helpful as I start taking a look at this.

I'm going to deep-dive the code tonight and put together some suggestions/requests, but I have a few notes from the surface-level stuff:

```
Blocks in Front Door
```
I'll give you an even better reason to exclude these, which is that there's no guaranteed that a given room will even exist in the seed, depending on the player's settings, and I wouldn't want to go through the effort of accounting for which rooms exist or not.


```
Checks for Invisible 1-up Checkpoints
```
I like this feature, but I think we need a different term for it than "Checkpoints", as that will be very often confused with the Midway Gates (I don't know if I've ever heard someone refer to them as Midway Gates instead of Checkpoints). I don't have a great alternative to suggest, maybe "Hidden 1-Ups"?


```
They're all coins (+1, +5, +10, + 15)
```
One might argue that finding someone's 1 coin is even worse than finding them a 1-Up, but it's at least variety, so I think it's fine. Could we change the 15 to 20 though? It's not a big deal, just feels like a better number to use.


```
Made the camera's vertical scroll edit optional
```
I have some concerns about this. I recognize that the existing solution isn't ideal, and a few people have complained about it, but having that turned off causes some levels to not be beatable where the logic says they should be, potentially leading to unbeatable seeds. Valley of Bowser 4 with only Yoshi and no Climb is an example of this. I'm certainly open to improvements to it, but I don't think allowing players to have that feature turned off, even voluntarily, is a great idea.


```
Added the possibility of swapping the player character with a custom one
```
This is interesting. Can you give a more detailed explanation of the process one would take to use this system?


As another consideration, how interested are you in updating the PopTracker pack in support of these new features? Should push come to shove, I can certainly do it, but tracker work is not my favorite activity.


Overall this looks fantastic. I'm very excited to give it all a try.

PoryGone commented on worlds/smw/Aesthetics.py:
```diff
@@ -196,22 +651,423 @@ def generate_shuffled_header_data(rom, world, player):
+            tileset = tileset_names[0x19]
+        palette = world.per_slot_randoms[player].randint(0, len(tilesets[tileset])-1)
+        rom.write_bytes(PALETTE_INDEX_ADDR + level_id, bytearray([palette]))
+        #print (f"Level: {level_id:03X}")
```
Please clean up the commented out prints here and throughout the rest of this file.

TheLX5 replied:
Done as well.

PoryGone commented on worlds/smw/Aesthetics.py:
```diff
@@ -196,22 +651,423 @@ def generate_shuffled_header_data(rom, world, player):
+        colors += bytearray([palette_file[(0xE1*2)+(i)] for i in range(14)])
+        colors += bytearray([palette_file[(0xF1*2)+(i)] for i in range(14)])
+
+    return colors
```
Please make sure there is a new line at the end of the file (git likes it better that way)

TheLX5 replied:
Done.

PoryGone commented on worlds/smw/Options.py:
```diff
@@ -109,6 +140,12 @@ class LevelShuffle(Toggle):
     display_name = "Level Shuffle"
 
 
+class EasyEarlyLevels(Toggle):
```
This shouldn't be an option. The primary motivator for the behaviour of having "easy" levels early isn't actually for the sake of the difficulty curve, but for making sure that we don't end up with too many levels with higher item requirements early in the seed. I'd like to preserve the old behaviour here.

TheLX5 replied:
We discussed this on discord and agree with you. I already reverted this addition.

PoryGone commented on worlds/smw/Options.py:
```diff
@@ -147,6 +184,29 @@ class DisplayReceivedItemPopups(Choice):
     option_progression = 2
     default = 2
 
+class BaseJunkWeight(Choice):
```
I don't know if this really needs to exist, to be honest. It makes sense for the traps, but I don't think coins vs 1-Ups for junk items is a sufficiently meaningful difference to have it take up option space.

TheLX5 replied:
Fair. Removed them and settled with static weights.

PoryGone commented on worlds/smw/Options.py:
```diff
@@ -204,6 +264,13 @@ class Autosave(DefaultOnToggle):
     display_name = "Autosave"
 
 
+class AdjustVerticalScroll(DefaultOnToggle):
```
As I mentioned in my other comment, I think we need to workshop an alternative solution to this problem.

TheLX5 replied:
While I've added the technology to make this setting per-room. I have not configured the rooms themselves. It is functional though.

PoryGone commented on worlds/smw/Levels.py:
```diff
@@ -261,7 +261,6 @@ def __init__(self, levelName: str, levelIDAddress: int, eventIDValue: int, exit1
     0x3D,
     0x01,
     0x0C,
-    0x0D,
```
I'd rather not reduce the size of the pool of levels for early game any further. This one is borderline, but the fewer levels we have here, the less variety we have early game.

PoryGone commented on worlds/smw/Levels.py:
```diff
@@ -493,6 +518,668 @@ def __init__(self, levelName: str, levelIDAddress: int, eventIDValue: int, exit1
     LocationName.special_zone_8_exit_1: [0x49, 0],
     LocationName.special_zone_8_dragon: [0x49, 2],
+
+
```
Trim some of the empty lines here please (one or two is sufficient)

TheLX5 replied:
I think I trimmed most of my whitespaces at this point. Wouldn't hurt to double check again.

PoryGone commented on worlds/smw/Levels.py:
```diff
@@ -528,39 +1215,77 @@ def generate_level_list(world, player):
-    if not world.exclude_special_zone[player]:
-        single_levels_copy.extend(special_zone_levels_copy)
-    world.random.shuffle(single_levels_copy)
+    if world.easy_early_levels[player]:
```
Please see my later comment in `Options.py` about why I would like this addition reverted.

PoryGone commented on worlds/smw/Locations.py:
```diff
@@ -234,6 +930,123 @@ def __init__(self, player: int, name: str = '', address: int = None, parent=None
+]
+
+special_zone_blocksanity_names = [
+    LocationName.special_zone_8_yoshi_block_1,
```
It's not a huge deal, but is there any reason in particular these are not listed in order?

TheLX5 replied:
They were generated with the level ID and I just copied them from the table above (`blocksanity_location_table`). I could rearrange them but doesn't seem super necessary I think?

PoryGone commented on worlds/smw/Names/LocationName.py:
```diff
@@ -4,12 +4,15 @@
 
 yoshis_island_1_exit_1  = "Yoshi's Island 1 - Normal Exit"
 yoshis_island_1_dragon  = "Yoshi's Island 1 - Dragon Coins"
+yoshis_island_1_moon    = "Yoshi's Island 1 - 3up Moon"
```
Slightly pedantic, but the official usages are denoted as "1-Up" and "3-Up". We should follow that style, since we've done so thus far.

TheLX5 replied:
Done.

PoryGone commented on worlds/smw/Names/LocationName.py:
```diff
@@ -4,12 +4,15 @@
+yoshis_island_3_bonus_block  = "Yoshi's Island 3 - 1up from Bonus Block"
 yoshis_island_4_exit_1  = "Yoshi's Island 4 - Normal Exit"
 yoshis_island_4_dragon  = "Yoshi's Island 4 - Dragon Coins"
+yoshis_island_4_checkpoint  = "Yoshi's Island 4 - Invisible 1up Checkpoints"
```
Mentioned in another comment in `Options.py`, but we should probably come up with a less confusing term for these than "Checkpoints"

TheLX5 replied:
We settled with "Hidden 1-Up" and made the corresponding change.

PoryGone commented on worlds/smw/Regions.py:
```diff
@@ -1042,13 +2232,16 @@ def connect_regions(world, player, level_to_tile_dict):
             lambda state: state.has(ItemName.p_switch, player))
     connect(world, player, names, LocationName.star_road_5_region, LocationName.star_road_5_exit_2,
-            lambda state: (state.has(ItemName.mario_carry, player) and
+            lambda state: ((state.has(ItemName.mario_carry, player) and
```
I'd rather not make this an exception to the "No Yoshi Fly" in logic rule that I've held thus far. It's fine for that to be a sequence break imo

PoryGone commented on worlds/smw/__init__.py:
```diff
@@ -40,6 +42,7 @@ class SMWWorld(World):
     """
     game: str = "Super Mario World"
     option_definitions = smw_options
+    #settings: typing.ClassVar[SMWSettings]
```
Why is this commented out?

TheLX5 replied:
Back when I started developing this (AP 0.4.1), I grabbed the py files from AP's source and my AP (installed version) was yelling at me due to not being able to use `import settings`, so I ended up commenting out the settings lines from `__init__.py` to get it to work.

TheLX5 replied:
Seems that it works with AP 0.4.3. I added it back.

PoryGone reviewed (changes requested):
I've covered just about everything, aside from an in-depth analysis of the logic for the new locations.

TheLX5 commented:
> ```
> Blocks in Front Door
> ```
> 
> I'll give you an even better reason to exclude these, which is that there's no guaranteed that a given room will even exist in the seed, depending on the player's settings, and I wouldn't want to go through the effort of accounting for which rooms exist or not.

This reason is much better haha. Agreed. I always play Yoshi Egg hunt so I never enter Bowser's Castle

> ```
> Checks for Invisible 1-up Checkpoints
> ```
> 
> I like this feature, but I think we need a different term for it than "Checkpoints", as that will be very often confused with the Midway Gates (I don't know if I've ever heard someone refer to them as Midway Gates instead of Checkpoints). I don't have a great alternative to suggest, maybe "Hidden 1-Ups"?

Might be good to change it, I named them like that due to Lunar Magic naming convention, but since this is accesible to people outside of the SMW modding community I'd go along with anything you suggest. Hidden 1-Ups sounds good so far. We could host a poll or something at the public discord channel.

> ```
> They're all coins (+1, +5, +10, + 15)
> ```
> 
> One might argue that finding someone's 1 coin is even worse than finding them a 1-Up, but it's at least variety, so I think it's fine. Could we change the 15 to 20 though? It's not a big deal, just feels like a better number to use.

We can, I'd advise to not use a big variety of numbers due to VRAM concerns. We could go with something even higher like +50 coins and making it rarer in the process.

> ```
> Made the camera's vertical scroll edit optional
> ```
> 
> I have some concerns about this. I recognize that the existing solution isn't ideal, and a few people have complained about it, but having that turned off causes some levels to not be beatable where the logic says they should be, potentially leading to unbeatable seeds. Valley of Bowser 4 with only Yoshi and no Climb is an example of this. I'm certainly open to improvements to it, but I don't think allowing players to have that feature turned off, even voluntarily, is a great idea.

Ah, this is a valid opinion. I could bake in a per-level override that's not accesible to the end user where we enable or disable this feature. I already placed a piece of code that places the current level ID in RAM (which was only used internally and never stored) that's being currently used for the palette randomizer. The code is super simple and the only "concern" would be that this option requires a 512 byte table, but that's negligible considering we have a 1MB game now.

> ```
> Added the possibility of swapping the player character with a custom one
> ```
> 
> This is interesting. Can you give a more detailed explanation of the process one would take to use this system?

The graphics were decompressed from the game itself (no copyrighted files included) to the expanded area of the game. This makes the replacement much easier similar to how ALTTP does it. The graphics at this moment are at the PC offset 0xE0000 and the auxiliary tiles are at 0xE6000, Yoshi (and other animated tiles) can be replaced and they're located at 0xE8800. If you want to see it by yourself, generate a game in AP 0.4.3 using the files I uploaded and open the patched game in a tool that can render SNES graphics (YY-CHR), you should be able to see all of these graphics lying there ready to be replaced. 
This also frees up 22000 bytes of RAM at $7E2000... but we don't use RAM here so it really doesn't matter (not to mention that SMW has a bunch of RAM that goes unused at bank $7F.

> As another consideration, how interested are you in updating the PopTracker pack in support of these new features? Should push come to shove, I can certainly do it, but tracker work is not my favorite activity.

Never programmed for PopTracker, depending on the complexity I might be able to tinker around with the current one and produce something useable.

> Overall this looks fantastic. I'm very excited to give it all a try.

Thanks!

PoryGone commented:
> > One might argue that finding someone's 1 coin is even worse than finding them a 1-Up, but it's at least variety, so I think it's fine. Could we change the 15 to 20 though? It's not a big deal, just feels like a better number to use.
> 
> We can, I'd advise to not use a big variety of numbers due to VRAM concerns. We could go with something even higher like +50 coins and making it rarer in the process.

I think either 20 or 50 would be a fine change to the 15, whichever you prefer.

> > I have some concerns about this. I recognize that the existing solution isn't ideal, and a few people have complained about it, but having that turned off causes some levels to not be beatable where the logic says they should be, potentially leading to unbeatable seeds. Valley of Bowser 4 with only Yoshi and no Climb is an example of this. I'm certainly open to improvements to it, but I don't think allowing players to have that feature turned off, even voluntarily, is a great idea.
> 
> Ah, this is a valid opinion. I could bake in a per-level override that's not accesible to the end user where we enable or disable this feature. I already placed a piece of code that places the current level ID in RAM (which was only used internally and never stored) that's being currently used for the palette randomizer. The code is super simple and the only "concern" would be that this option requires a 512 byte table, but that's negligible considering we have a 1MB game now.
> 
I think if we can identify the levels that have issues with the camera lock (I just remembered there are some Dragon Coins that can't be collected off-screen if the camera is vertical-locked, so that's another case), that would be a fine in-between solution, to  at least minimize the amount of levels in which people have to deal with the weird unlocked camera. If what you described is reasonable, I'll let you try cooking on that.

> 
> > As another consideration, how interested are you in updating the PopTracker pack in support of these new features? Should push come to shove, I can certainly do it, but tracker work is not my favorite activity.
> 
> Never programmed for PopTracker, depending on the complexity I might be able to tinker around with the current one and produce something useable.
> 
Not a huge deal either way, but fwiw, the tracker pack is like 90% JSON and 10% Lua (which I can certainly handle the Lua side). 
If you take a look and think you might want to take a crack at it, we can chat about how I'd envision it looking. I just dread the concept of manually copying over like 600 locations and their associated logic. Mayhaps you'll see a way to autogenerate the tracker logic data in much the same way as you did for the generation logic.

TheLX5 committed:
Removed code used for debugging

TheLX5 committed:
Reverted EasyEarlyLevels option

TheLX5 committed:
Location name adjustments

TheLX5 committed:
Reverted added logic for Funky

TheLX5 committed:
Removed easy early levels option

TheLX5 committed:
Removed options for junk item weights

TheLX5 committed:
Reworked the vertical scroll hack

TheLX5 committed:
Change item names from the init file

TheLX5 committed:
trimming

TheLX5 commented:
I've handled everything except the conflict, the coin amount change and the commented out lines in `__init__.py` regarding the settings.

I'm commenting here now since I need a bit guidance. I've already commited the necessary changes to the camera's vertical scroll to work in a per-room basis (https://github.com/PoryGoneDev/Pory_Archipelago/pull/18/commits/3fad0ef4c7b558d176178d796ca782e95365eea4). I'd just need to know which rooms (or whole levels) would you like to have a less restrictive camera and I'll configure the table by myself.

TheLX5 committed:
Blocksanity data is now computed from the json file instead of using a bin file

TheLX5 committed:
Added a missing blocksanity location

TheLX5 committed:
Bring back import settings and related code

TheLX5 committed:
Bring back more settings... again!

TheLX5 committed:
Attempt to solve merge conflict

TheLX5 committed:
Bring back ROM expansion

TheLX5 committed:
Attempt #2 to fix merge conflicts

TheLX5 committed:
Changed junk item "15 coins" to "50 coins"

TheLX5 commented:
I think everything's covered now? Dunno if I should click the resolve button yet on those requested changes. 

unrelated: 
lol, forgot GitHub does this
![image](https://github.com/PoryGoneDev/Pory_Archipelago/assets/17969056/b12f4ade-0db9-4f4a-a664-787dad7abd15)

TheLX5 committed:
Remove unused things

TheLX5 committed:
Configured the vertical scroll setting table

TheLX5 committed:
Modified the map indicator for Hidden 1-Ups

TheLX5 committed:
Adjusted some palettes

TheLX5 commented:
Added a few things that I missed. Also configured the vertical scroll table.

Activated in:
* Yoshi's Island 2 (Main room)
* Donut Plains 2 (Key room)
* Donut Plains 3 (Main room)
  * Dragon coin above jumping koopas seems reachable, but different camera scrolls might hinder the level experience. Needs testing
* Donut Ghost House (Secondary rooms)
* Donut Secret House (Blue doors rooms)
* Donut Secret 2 (Main room)
* Vanilla Dome 2 (Main room)
* Vanilla Ghost House (Main room & its copy)
* Vanilla Secret 1 (Main room)
* Ludwig castle (Main room)
* Chocolate Island 5 (Main room)
* Valley of Bowser 4 (Main room)

Unsure (not activated):
* Yoshi's Island 1
  * Moon is too high in the sky to reach without flying
* Donut Plains 1 (Main room)
  * Dragon coin above Chucks seems to high to reach without flying
* Donut Plains 4 (Main room)
  * 3-Up Moon & Hidden 1-Up are too high in the sky to reach without flying
* Butter Bridge 2
  * Dragon coin in the sky seems too high to reach without flying
* Cookie Mountain (Main room)
  * Dragon coin after checkpoint seems to high to reach without flying
* Forest of Illusion 4
  * Dragon coin above the second Lakitu seems too high in the sky to reach without flying with cape or a cloud (needs carry)
* Chocolate Island 1 (Copy of main room)
  * Moon is too high in the sky to reach without flying
* Star World 5
  * Doesn't feel necessary as you would rarely go too high unless you're attempting to get the secret exit but then you'd never fall down again unless it's an accident

Refused to activate:
* Vanilla Dome 1
  * Doesn't seem too necessary here
* Vanilla Dome 3
  * Level has a really annoying blind jump when the camera setting is changed
* Forest Ghost House
  * Doesn't seem too necessary here
* Chocolate Island 4
  * Level flows much better with the original camera setting imo, especially in the large lava pit
* Star World 3
  * Doesn't need it
* Funky
  * Doesn't need it

TheLX5 committed:
Changed 1-Up Checkpoint references to Hidden 1-Up

Also changed some of the options' descriptions
... and changed the name of a location in Donut Ghost House

TheLX5 committed:
Fixed the logic for a block that could be either a green or yellow switch palace block

TheLX5 committed:
Make impossible to receive checks during boss battles

TheLX5 commented:
Added some small changes:
* A block in Donut Ghost House was renamed as it (either the Vine or the P-Switch block) fails to send a check. Since my blocksanity implementation doesn't take into account room ID, only block coordinates, only one of those will send a check. 
* A Switch Palace block in Chocolate Island 1 was renamed and its logic was adjusted as it can be either Green or Yellow depending on the sub room you're in because for some reason that level sends you to multiple copies of itself via warp pipes, and only some of them contain a 3-Up Moon and the Switch Palace block gets swapped with another color.
* Made impossible to receive checks during boss battles. It broke my in-level indicators of newly received items.
* Changed every reference to 1-Up Checkpoints in my code, they're now Hidden 1-Ups.

There's some minor visual issues with the Curated option for the palette shuffle:
* When you eat a berry with Yoshi, in some palette configurations, the berry will leave a green bush tile behind instead of the actual bush color (white in some snow themed palettes).
* Some palettes that weren't properly configured in the cave theme
* Some palettes didn't properly paint their bushes tiles with berries

I'll try to tinker around to get those fixed, will need some extra ASM code but it'll be worth.

TheLX5 committed:
Several palette fixes

arbustos

TheLX5 commented:
And done. I think that's everything fixed/adjusted.

TheLX5 committed:
Rework palette shuffle options

TheLX5 commented:
Seems that I forgot to remove an option. Also changed the palette options according to our Discord chat.

PoryGone commented on worlds/smw/Aesthetics.py:
```diff
@@ -160,21 +614,21 @@ def generate_shuffled_ow_music(world, player):
 
 def generate_shuffled_header_data(rom, world, player):
-    if world.music_shuffle[player] != "full" and not world.foreground_palette_shuffle[player] and not world.background_palette_shuffle[player]:
+    if world.music_shuffle[player] != "full" and world.foreground_palette_shuffle[player] == "on_legacy" and world.background_palette_shuffle[player] == "on_legacy" :
```
We should early-out here when the shuffles are off, right? So maybe something like

```suggestion
    if world.music_shuffle[player] != "full" and world.foreground_palette_shuffle[player] == "off" and world.background_palette_shuffle[player] == "off" :
```

PoryGone replied:
Nvm, restructuring this following Discord conversation.
